core/src/main/java/org/elasticsearch/index/mapper/internal/ParentFieldMapper.java

```diff
@@ -223,20 +236,36 @@ public Query termsQuery(List values, @Nullable QueryParseContext context) {
     }
 
     private final String type;
+    private final MappedFieldTypeReference joinFieldTypeRef;
+    private final MappedFieldType defaultJoinFieldType;
 
-    protected ParentFieldMapper(MappedFieldType fieldType, String type, Settings indexSettings) {
-        super(NAME, setupDocValues(indexSettings, fieldType), setupDocValues(indexSettings, Defaults.FIELD_TYPE), indexSettings);
+    protected ParentFieldMapper(MappedFieldType fieldType, MappedFieldType joinFieldType, String type, Settings indexSettings) {
+        super(NAME, fieldType, Defaults.FIELD_TYPE, indexSettings);
         this.type = type;
+        this.joinFieldTypeRef = new MappedFieldTypeReference(joinFieldType);
+        this.defaultJoinFieldType = Defaults.JOIN_FIELD_TYPE;
+    }
+
+    public ParentFieldMapper(Settings indexSettings, MappedFieldType existing, String parentType) {
+        this(existing == null ? Defaults.FIELD_TYPE.clone() : existing.clone(), joinFieldTypeForParentType(parentType, indexSettings), null, indexSettings);
     }
 
-    public ParentFieldMapper(Settings indexSettings, MappedFieldType existing) {
-        this(existing == null ? Defaults.FIELD_TYPE.clone() : existing.clone(), null, indexSettings);
+    private static MappedFieldType joinFieldTypeForParentType(String parentType, Settings indexSettings) {
+        MappedFieldType joinFieldType = Defaults.JOIN_FIELD_TYPE.clone();
+        joinFieldType.setNames(new MappedFieldType.Names(joinField(parentType)));
+        joinFieldType.setFieldDataType(new FieldDataType("_parent", settingsBuilder().put(MappedFieldType.Loading.KEY, MappedFieldType.Loading.LAZY)));
+
+        Version indexCreated = Version.indexCreated(indexSettings);
+        if (indexCreated.before(Version.V_2_0_0_beta1)) {
+            joinFieldType.setHasDocValues(false);
+        }
+
+        return joinFieldType;
     }
 
-    static MappedFieldType setupDocValues(Settings indexSettings, MappedFieldType fieldType) {
-        fieldType = fieldType.clone();
-        fieldType.setHasDocValues(Version.indexCreated(indexSettings).onOrAfter(Version.V_2_0_0_beta1));
-        return fieldType;
+    @Override
+    public FieldDataType fieldDataType() {
```

Comment:
The more I think about this, the more I still think having something like `getJoinFieldType()` is the way to go. Then there is no need for `joinField` below, because the field name is included with the field type?